### DISCUSSION TITLE
How to add conic constraints?

### DISCUSSION THREAD

**Discussion by imuncle** [TinyMPC/TinyMPC#23] [Q&A]
Hi

In the introduction, I found that TinyMPC is claimed to support conic constraints. But I can only find the setup API of `A, B, Q, and R` matrices, and the upper/lower bound of `x` and `u` :

```cpp
int tiny_setup(TinySolver** solverp,
                tinyMatrix Adyn, tinyMatrix Bdyn, tinyMatrix Q, tinyMatrix R, 
                tinytype rho, int nx, int nu, int N,
                tinyMatrix x_min, tinyMatrix x_max, tinyMatrix u_min, tinyMatrix u_max,
                int verbose);
```

Is there any API available for conic constraints?

Thank you for your time and efforts

**Comment by xkhainguyen** [comment]
Hi there,

Thank you for your question. Support for second-order conic constraints is provided in [our recent submission](https://arxiv.org/abs/2403.18149) to CDC 2024. The Crazyflie firmware accompanying the paper can be found [here](https://github.com/RoboticExplorationLab/tinympc-crazyflie-firmware/tree/main/examples/controller_tinympc_state_soc). We created a new temporary TinyMPC2 library.
However, the API is still under private development to improve UI. If you need access immediately, please feel free to contact us via email.

Stay tuned for public updates coming very soon.
